Add kn config file docs and section for plugins (#3495)
knative/docs: docs/client/configure-kn.md

---
title: "Customizing kn"
weight: 03
type: "docs"
---

> You can customize your `kn` CLI setup by creating a `config.yaml` configuration file. The CLI will look for this file under the home directory of the user at `$HOME/.config/kn/config.yaml`, however you can create this file anywhere and use the `--config` flag to specify its path.

Reviewer: ```suggestion
You can customize your `kn` CLI setup by creating a `config.yaml` configuration file. The configuration can be either provided via a `--config` flag or is picked up from a default location. The default configuration location conforms to the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) and is different for Unix systems and Windows systems. If `XDG_CONFIG_HOME` env var is set, the default config location `kn` looks for is `$XDG_CONFIG_HOME/kn`, otherwise file is looked up under the home directory of the user at `$HOME/.config/kn/config.yaml`. For Windows systems, the default `kn` configuration location is `$APPDATA\kn`.
```

## Example configuration file

```yaml

> lookupPluginsInPath: true
> pluginsdir: ~/.config/kn/plugins
> sink:
> - prefix: myprefix
>   group: eventing.knative.dev
>   version: v1alpha1
>   resource: brokers

Reviewer 2: ```suggestion
plugins:
  pathLookup: true
  directory: ~/.config/kn/plugins
eventing:
  sink:
  - prefix: svc
    group: core
    version: v1
    resource: services
```

Apologies, that the original documentation is not up to date (going to send a PR to fix https://github.com/knative/client/tree/main/docs#kn-config).

(also I updated the sample mapping to a more realistic example, as this could be used directly to map Kubernetes Services to a `svc:` prefixes (while Knative Services are mapped to `ksvc:` already by default)

This means, we have introduced a grouping in the configuration on the first level (`plugins` and `eventing`), this might deserve separate sub-sections maybe in the documentation ?

Author: Made some updates. I think giving them separate sections for now might be overkill tbh until we get to adding more complex docs about configuration for `kn`. IMO for now it's enough to explain them in one example that can be iterated later.

```

Where

> - `lookupPluginsInPath` specifies whether `kn` should look for [plugins](./kn-plugins) in the specified `PATH` environment variable. This is a boolean configuration option. The default value is `false`.

Author: removed "specified", thanks for clarifying

- `pluginsdir` specifies the directory where `kn` will look for plugins. The defaults path is `~/.config/kn/plugins`. This can be any directory that is visible to the user.

- `sink` defines the prefix that is used to refer to Kubernetes Addressable resources. To configure a sink prefix, you must define following fields, as shown in the example:

>     - `prefix`: The prefix you want to use to describe your sink. Service, `svc`, and `broker` are predefined prefixes in `kn`.

Reviewer 2: Could be that the original documentation out of sync, but the currently predefined prefixes are:

* `ksvc` for a Knative service
* `broker` for a broker
* `channel` for a channel.

---
title: "Customizing kn"
weight: 03
type: "docs"
---

You can customize your `kn` CLI setup by creating a `config.yaml` configuration file. You can provide this configuration by using the `--config` flag, otherwise the configuration is picked up from a default location. The default configuration location conforms to the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html), and is different for Unix systems and Windows systems.

- If the `XDG_CONFIG_HOME` environment variable is set, the default configuration location that `kn` looks for is `$XDG_CONFIG_HOME/kn`.
- If the `XDG_CONFIG_HOME` environment variable is not set, `kn` looks for the configuration in the home directory of the user at `$HOME/.config/kn/config.yaml`.
- For Windows systems, the default `kn` configuration location is `%APPDATA%\kn`.

## Example configuration file

```yaml
plugins:

>   pathLookup: true

Reviewer 2: ```suggestion
  path-lookup: true
```

  directory: ~/.config/kn/plugins
eventing:

>   sink:

Reviewer 2: ```suggestion
  sink-mappings:
```

  - prefix: svc
    group: core
    version: v1
    resource: services
```

Where

> - `pathLookup` specifies whether `kn` should look for [plugins](./kn-plugins) in the `PATH` environment variable. This is a boolean configuration option. The default value is `false`.

Reviewer 2: ```suggestion
- `path-lookup` specifies whether `kn` should look for [plugins](./kn-plugins) in the `PATH` environment variable. This is a boolean configuration option. The default value is `false`.
```

- `directory` specifies the directory where `kn` will look for plugins. The default path depends on the operating system, as described above. This can be any directory that is visible to the user.

> - `sink` defines the Kubernetes Addressable resource that is used when you use the `--sink` flag with a `kn` CLI command.

Author: ```suggestion
- `sink-mappings` defines the Kubernetes Addressable resource that is used when you use the `--sink` flag with a `kn` CLI command.
```
